### PR TITLE
Infer types in default expression

### LINKED ISSUE #6765: Completion: preselection on default(
Paste C#:
```C#
class C
{
    static void Main(string[] args)
    {
        System.ConsoleModifiers c = $$
    }
}
```

Remove the ```$$``` and type ```default(```

**Expected**: Completion list shows, preselected to ```System.ConsoleModifiers```
**Actual**: No completion list

### PR DESCRIPTION
**Pull request by rchande**
Fixes #6765.

### CODE REVIEW

**Comment by rchande** [comment]
Tagging @dotnet/roslyn-ide for review.

**Comment by davkean** [comment]
Github timed out.

retest prtest/mac/dbg/unit32 please

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs [hunk 1]
```diff
@@ -1035,6 +1036,11 @@ private IEnumerable<ITypeSymbol> InferTypeInConditionalExpression(ConditionalExp
                            : otherTypes;
             }
 
+            private IEnumerable<ITypeSymbol> InferTypeInDefaultExpression(DefaultExpressionSyntax defaultStatement)
+            {
+                return InferTypes(defaultStatement.SpanStart);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs]: InferTypes(defaultStatement.parent)

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs [hunk 2]
```diff
@@ -1035,6 +1036,11 @@ private IEnumerable<ITypeSymbol> InferTypeInConditionalExpression(ConditionalExp
                            : otherTypes;
             }
 
+            private IEnumerable<ITypeSymbol> InferTypeInDefaultExpression(DefaultExpressionSyntax defaultStatement)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs]: "defaultExpression" (as opposed to Statement).

File: src/EditorFeatures/CSharpTest/TypeInferrer/TypeInferrerTests.cs [hunk 3]
```diff
@@ -1761,5 +1761,35 @@ public class C
...
+        public void TestDefaultStatement2()
+        {
+            var text =
+    @"class C
+{
+    static void Foo(System.ConsoleModifiers arg)
+    {
+        Foo(default([||])
+    }
+}";
```
Comment (jmarolf) [inline src/EditorFeatures/CSharpTest/TypeInferrer/TypeInferrerTests.cs]: What happens if default() is used to assign a field outside of a method?

**Comment by rchande** [comment]
@CyrusNajmabadi Implemented your feedback
@jmarolf Our thoroughly tested code to infer in EqualsValueClauses runs.

**Comment by jmarolf** [comment]
:+1:

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs [hunk 4]
```diff
@@ -204,6 +204,7 @@ protected override IEnumerable<ITypeSymbol> InferTypesWorker_DoNotCallDirectly(i
                     (CatchFilterClauseSyntax catchFilterClause) => InferTypeInCatchFilterClause(catchFilterClause, token),
                     (CheckedExpressionSyntax checkedExpression) => InferTypes(checkedExpression),
                     (ConditionalExpressionSyntax conditionalExpression) => InferTypeInConditionalExpression(conditionalExpression, previousToken: token),
+                    (DefaultExpressionSyntax defaultStatement) => InferTypeInDefaultExpression(defaultStatement),
```
Comment (jasonmalinowski) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs]: s/Statement/Expression/

**Comment by davkean** [comment]
@CyrusNajmabadi @Pilchie @jasonmalinowski @DustinCampbell Can one of you sign off on this please?

**Comment by CyrusNajmabadi** [comment]
:+1:

File: src/EditorFeatures/CSharpTest/TypeInferrer/TypeInferrerTests.cs [hunk 5]
```diff
@@ -1761,5 +1761,35 @@ public class C
 }";
             Test(text, "System.Int32");
         }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.TypeInferenceService)]
+        [WorkItem(6765, "https://github.com/dotnet/roslyn/issues/6765")]
+        public void TestDefaultStatement1()
```
Comment (DustinCampbell) [inline src/EditorFeatures/CSharpTest/TypeInferrer/TypeInferrerTests.cs]: `TestDefaultExpression1`

File: src/EditorFeatures/CSharpTest/TypeInferrer/TypeInferrerTests.cs [hunk 6]
```diff
@@ -1761,5 +1761,35 @@ public class C
...
+    {
+        System.ConsoleModifiers c = default([||])
+    }
+}";
+            Test(text, "global::System.ConsoleModifiers", testNode:false);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.TypeInferenceService)]
+        [WorkItem(6765, "https://github.com/dotnet/roslyn/issues/6765")]
+        public void TestDefaultStatement2()
```
Comment (DustinCampbell) [inline src/EditorFeatures/CSharpTest/TypeInferrer/TypeInferrerTests.cs]: `TestDefaultExpression2`

**Comment by DustinCampbell** [comment]
:+1:

**Comment by Pilchie** [comment]
:+1:

**Comment by Pilchie** [comment]
Turns out this also fixes #1647.  (Sorry @slaks for us forgetting about your original report).

**Pull request merged**